src/environmentd/src/deployment/preflight.rs

```diff
@@ -237,25 +241,76 @@ async fn check_ddl_changes(
     initial_next_user_item_id: u64,
     initial_next_replica_id: u64,
 ) {
-    let (next_user_item_id, next_replica_id) = get_next_ids(
-        boot_ts,
-        persist_client.clone(),
-        environment_id.clone(),
-        deploy_generation,
-        Arc::clone(&catalog_metrics),
-        bootstrap_args.clone(),
+    let openable_adapter_storage = mz_catalog::durable::persist_backed_catalog_state(
+        persist_client,
+        environment_id.organization_id(),
+        BUILD_INFO.semver_version(),
+        Some(deploy_generation),
+        catalog_metrics,
     )
-    .await;
+    .await
+    .expect("incompatible catalog/persist version");
+
+    let (mut catalog, _audit_logs) = openable_adapter_storage
+        .open_savepoint(boot_ts, &bootstrap_args)
+        .await
+        .unwrap_or_terminate("can open in savepoint mode");
+
+    let tx = catalog
+        .transaction()
+        .await
+        .unwrap_or_terminate("unexpected error while getting transaction");
+
+    // We must explicitly check the catalog for these IDs since IDs can be
+    // allocated during sequencing/planning but not yet committed to the catalog.
+    let new_replicas = tx
+        .get_cluster_replicas()
+        .filter_map(|replica| match replica.replica_id {
+            ReplicaId::User(id) if id >= initial_next_replica_id => Some(replica),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    let new_objects = tx
+        .get_items()
+        .filter_map(|item| match item.id {
+            CatalogItemId::User(id) if id >= initial_next_user_item_id => Some(item),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
 
     tracing::info!(
-        %initial_next_user_item_id,
-        %initial_next_replica_id,
-        %next_user_item_id,
-        %next_replica_id,
-        "checking if there was any relevant DDL");
-
-    if next_user_item_id > initial_next_user_item_id || next_replica_id > initial_next_replica_id {
-        halt!("there have been DDL that we need to react to; rebooting in read-only mode")
+        potential_new_user_item_id=%initial_next_user_item_id,
+        potential_new_replica_id=%initial_next_replica_id,
+        "checking if there have been new IDs committed to the catalog since the DDL check began;"
+    );
```

Comment:
I think this `info` doesn't add much value anymore, since we're already logging `initial_next_{user_item,replica}_id` further up ("waiting for deployment to be caught up"). Maybe it's useful to just log the "checking if there was any relevant DDL" line without the additional fields, so we still can verify that the check runs from time to time.